### PR TITLE
[CHE-170] Refactor Global Error Handler

### PR DESCRIPTION
**Pull request by seantokuzo**
### Description

As Part of the CHE-167 story to refactor error handling, this PR aims to:

Create and use a global error handler middleware that handles:
  - our custom error classes
  - errors thrown the old way (temporary until middlewares/controllers can be refactored to use custom error classes)
  - unknown errors by sending back a generic error response

The goal of this error handler is:
- to consistently format error responses sent back to the client (with the help of our custom error class `serializeErrors` method): `{ message: string, field?: string }[]`
- to set relevant status codes
- to log helpful information for debugging errors

This PR also:
- has the app use the new global error handler
- simplifies importing custom error classes from the errors folder (via `server/errors/index.ts`)

### Jira Task

[JIRA TICKET](https://codehammers.atlassian.net/browse/CHE-170?atlOrigin=eyJpIjoiZDhjYjYwOWViNjBhNDZmYmFmNzQ1YzE3NGIwOWQxNGEiLCJwIjoiaiJ9)

### Testing Instructions
(Make sure you have the most recent image by running `npm run docker-remove-all` first)

Check if current tests still pass with:
- `npm run docker-test:all`

Make sure error handler is working
- add `- 3000:3000` to the ports array in docker-compose-dev.yml
- `npm run docker-dev`
- on a browser navigate to unhandled route - eg. [http://localhost:3000/non-existent-route](http://localhost:3000/non-existent-route)
- expect our generic error message "Something went wrong, please try again later"

#### All Team Members

- [x] I added a descriptive title to this PR.
- [x] I filled out the **Description**, **Jira Task**, and **Testing Instructions** sections above.
- [ ] I added or updated [Jest unit tests]for any changes to components, server-side controllers, etc.
- [x] I ran `npm run docker-test` in my local environment to check that this PR passes all unit tests.
- [x] I did a quick check to make sure my code changes follow the recommended style guide.

### CODE REVIEW

File: server/middleware/errorHandler.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+  // Handle errors thrown the old way
+  if (!(err instanceof Error) && err.status && err.log && err.message) {
+    console.log(err.log);
+    return res.status(err.status).send([{ message: err.message.err }]);
+  }
+
+  // If it is an unknown error send back a generic error
+  const internalError = new InternalError();
+  return res.status(internalError.statusCode).send(internalError.serializeErrors());
```
Comment (brok3turtl3) [inline server/middleware/errorHandler.ts]: Great transitional logic here @seantokuzo !

File: server/errors/index.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+export * from './badRequestError';
+export * from './customError';
+export * from './databaseConnectionError';
+export * from './internalError';
+export * from './notAuthorizedError';
+export * from './notFoundError';
+export * from './requestValidationError';
+export * from './validationError';
```
Comment (brok3turtl3) [inline server/errors/index.ts:8]: Good clean up

**Review by brok3turtl3 — APPROVED** [review]
This is all looking good - Excited to see it roll out across the BE!

File: server/middleware/errorHandler.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+interface OldError {
+  status: number;
+  log: string;
+  message: { err: string };
+}
+
+const errorHandler = (err: Error | OldError, _req: Request, res: Response, _next: NextFunction) => {
+  // If it is one of our custom errors use its properties/method
+  if (err instanceof CustomError) {
+    console.log(err.message);
```
Comment (Mnelson98) [inline server/middleware/errorHandler.ts]: This looks great to me! I'm curious if 'console.error' would be more fitting here, however? Maybe not, just wondering.

Comment (seantokuzo) [reply server/middleware/errorHandler.ts]: Great catch! `console.error` is def more appropriate.

I think I also need to be logging the entire err to get the stack trace for debugging. Refactor coming in hot

**Review by Mnelson98 — APPROVED** [review]
Looks good to me!

**Review by jimpulsee — APPROVED** [review]
Good work!

**Comment by brok3turtl3** [comment]
Good job @seantokuzo!

**Pull request merged**